src/netCDF4/__init__.pyi

```diff
@@ -280,152 +409,168 @@ class Dataset:
     def __getattr__(self, name: str) -> Any: ...
     def __delattr__(self, name: str): ...
-    def __dealloc(self) -> None: ...
     def __reduce__(self) -> NoReturn: ...
     def __enter__(self) -> Self: ...
     def __exit__(self, atype, value, traceback) -> None: ...
 
-
 class Group(Dataset):
     def __init__(self, parent: Dataset, name: str, **kwargs: Any) -> None: ...
-
     def close(self) -> NoReturn: ...
 
-
 class Dimension:
     def __init__(self, grp: Dataset, name: str, size: int | None = None, **kwargs: Any) -> None: ...
-
     @property
     def name(self) -> str: ...
     @property
     def size(self) -> int: ...
-
     def group(self) -> Dataset: ...
     def isunlimited(self) -> bool: ...
-
     def __len__(self) -> int: ...
 
+class _VarDatatypeProperty:
+    # A faux descriptor definition of the property to allow overloads
+    @overload
+    def __get__(self, instance: Variable[RealVarT], owner: Any) -> RealVarT: ...
+    @overload
+    def __get__(self, instance: Variable[ComplexVarT], owner: Any) -> CompoundType: ...
+    @overload
+    def __get__(self, instance: Variable[str], owner: Any) -> VLType: ...
+    @overload
+    def __get__(
+        self, instance: Variable[Any], owner: Any
+    ) -> Any: ...  # actual return type np.dtype | CompoundType | VLType | EnumType
 
-class Variable(Generic[T_Datatype]):
+class _VarDtypeProperty:
+    # A faux descriptor definition of the property to allow overloads
+    @overload
+    def __get__(self, instance: Variable[NumericVarT], owner: Any) -> np.dtype[NumericVarT]: ...
+    @overload
+    def __get__(self, instance: Variable[str], owner: Any) -> type[str]: ...
+    @overload
+    def __get__(self, instance: Variable[Any], owner: Any) -> Any: ...  # actual return type np.dtype | Type[str]
 
+class Variable(Generic[VarT]):
+    # Overloads of __new__ are provided for some cases where the Variable's type may be statically inferred from the datatype arg
     @overload
-    def __new__(  # type: ignore
-        self,
+    def __new__(
+        cls,
         grp: Dataset,
         name: str,
-        datatype: T_DatatypeNC,
-        dimensions: DimensionsOptions = (),
-        compression: CompressionOptions | None = None,
+        datatype: np.dtype[NumericVarT] | type[NumericVarT],
+        dimensions: DimensionsSpecifier = (),
+        compression: CompressionType | None = None,
         zlib: bool = False,
-        complevel: CompressionLevelOptions | None = 4,
+        complevel: CompressionLevel | None = 4,
         shuffle: bool = True,
-        szip_coding: Literal['nn', 'ec'] = 'nn',
+        szip_coding: Literal["nn", "ec"] = "nn",
         szip_pixels_per_block: Literal[4, 8, 16, 32] = 8,
         blosc_shuffle: Literal[0, 1, 2] = 1,
         fletcher32: bool = False,
         contiguous: bool = False,
-        chunksizes: Sequence[int] | None = None,
-        endian: EndianOptions = 'native',
+        chunksizes: int | None = None,
```

Comment:
Are you sure that the sequence is not correct?

Reply:
Good catch. This was an erroneous change. I've reverted it.